components/modules/crypto.c

```diff
@@ -5,14 +5,18 @@
 #include "mbedtls/md.h"
 #include "module.h"
 #include "platform.h"
+#include "sodium_module.h"
 
 #define HASH_METATABLE "crypto.hasher"
 
+// Must not overlap with mbedtls_md_type_t
+#define SODIUM_BLAKE2B (-1)
+
 // algo_info_t describes a hashing algorithm and output size
 typedef struct {
     const char* name;
     const size_t size;
-    const mbedtls_md_type_t type;
+    const int type; // either a mbedtls_md_type_t or SODIUM_BLAKE2B
```

Comment:
You don't reckon it might be cleaner to have something like
```
typedef enum { CRYPTO_PROVIDER_MBEDTLS, CRYPTO_PROVIDER_SODIUM } crypto_provider_t;
typedef enum { SODIUM_ALGO_BLAKE2B } sodium_algo_t;
...
typedef struct {
  const char *name;
  const size_t size;
  const crypto_provider_t provider;
  union {
    const mbedtls_md_type_t mbedtls;
    const sodium_algo_t sodium;
  } type; // or unnamed union if the toolchain supports it
} algo_info_t;
```
It would be a bit more resistant to accidental overlap if something changes in the `mbedtls_md_type_t` values, which would appear as rather silent error (and silent errors are "teh wurst").

Reply:
I'm in two minds as to which approach is the more likely to introduce silent errors - some subsequent code forgetting to check `provider` as well as `type`, versus the possibility that Mbed TLS in future changes how it defines `mbedtls_md_type_t`...? I picked `-1` for `SODIUM_BLAKE2B` on the basis it was pretty unlikely that `mbedtls_md_type_t` would start using that value (not to mention that it would also need to start being used by crypto.c before it'd become a problem).

In the event of a tie-break I'd be inclined to just go for the simpler solution, which doesn't introduce an additional variable and a union :)

Reply:
Until such a time we have more crypto libs being linked in here, I can be convinced that the simpler solution is the preferred one :)